File: src/main/resources/coremods/world_preset_codec.js
function initializeCoreMod() {
    const Opcodes = Java.type('org.objectweb.asm.Opcodes');
    const InsnList = Java.type('org.objectweb.asm.tree.InsnList');
    const MethodInsnNode = Java.type('org.objectweb.asm.tree.MethodInsnNode');

    return {
        'modify_codec': {
            'target': {
                'type': 'METHOD',
                'class': 'net.minecraft.world.level.levelgen.presets.WorldPreset',
                'methodName': '<clinit>',
                'methodDesc': '()V'
            },
            'transformer': function (method) {
                const target = new InsnList();
                target.add(new MethodInsnNode(
                    Opcodes.INVOKESTATIC,
                    'de/melanx/skyblockbuilder/util/CoreUtil',
                    'augmentWorldPresetCodec',
                    '(Lcom/mojang/serialization/Codec;)Lcom/mojang/serialization/Codec;',
                    false
                ));
                for (let i = 0; i < method.instructions.size(); i++) {
                    const node = method.instructions.get(i);
                    if (node.getOpcode() === Opcodes.PUTSTATIC) {
                        const fieldNode = node;
                        if (fieldNode.owner === 'net/minecraft/world/level/levelgen/presets/WorldPreset' && fieldNode.name === 'DIRECT_CODEC') {
                            method.instructions.insertBefore(fieldNode, target);
                            return method;
                        }
                    }
                }
                throw new Error('Failed to patch WorldPreset.class');
            }
        }
    };
}

File: coremods/src/main/java/de/melanx/skyblockbuilder/coremods/WorldPresetCodec.java
package de.melanx.skyblockbuilder.coremods;

import net.neoforged.neoforgespi.transformation.ProcessorName;
import net.neoforged.neoforgespi.transformation.SimpleMethodProcessor;
import net.neoforged.neoforgespi.transformation.SimpleTransformationContext;
import org.objectweb.asm.Opcodes;
import org.objectweb.asm.tree.*;

import javax.annotation.Nonnull;
import java.util.Set;

public class WorldPresetCodec extends SimpleMethodProcessor {

    @Override
    public void transform(MethodNode method, SimpleTransformationContext context) {
        InsnList target = new InsnList();
        target.add(
                new MethodInsnNode(Opcodes.INVOKESTATIC,
                        "de/melanx/skyblockbuilder/util/CoreUtil",
                        "augmentWorldPresetCodec",
                        "(Lcom/mojang/serialization/Codec;)Lcom/mojang/serialization/Codec;",
                        false)
        );

        for (AbstractInsnNode insn : method.instructions) {
            if (insn.getOpcode() == Opcodes.PUTSTATIC) {
                FieldInsnNode field = (FieldInsnNode) insn;
                if (field.owner.equals("net/minecraft/world/level/levelgen/presets/WorldPreset") && field.name.equals("DIRECT_CODEC")) {
                    method.instructions.insertBefore(insn, target);
                    return;
                }
            }
        }

        throw new IllegalStateException("Could not find field DIRECT_CODEC");
    }

    @Nonnull
    @Override
    public Set<Target> targets() {
        return Set.of(
                new Target(
                        "net.minecraft.world.level.levelgen.presets.WorldPreset",
                        "<clinit>",
                        "()V"
                )
        );
    }

    @Override
    public ProcessorName name() {
        return new ProcessorName("skyblockbuilder", "world_preset_codec");
    }
}
